File: 3219-make-lexicographically-smallest-array-by-swapping-elements/3219-make-lexicographically-smallest-array-by-swapping-elements.go
package main

import "sort"

type Pair struct {
    value int
    index int
}

func lexicographicallySmallestArray(nums []int, limit int) []int {
    n := len(nums)
    help := make([]Pair, n)

    for i := 0; i < n; i++ {
        help[i] = Pair{nums[i], i}
    }

    sort.Slice(help, func(i, j int) bool {
        return help[i].value < help[j].value
    })

    res := make([]int, n)
    pos := []int{}

    s, e := 0, 0
    for e < n {
        pos = append(pos, help[e].index)
        e++

        if e == n || help[e].value-help[e-1].value > limit {
            sort.Ints(pos)
            for _, idx := range pos {
                res[idx] = help[s].value
                s++
            }
            pos = []int{}
        }
    }

    return res
}

File: 3219-make-lexicographically-smallest-array-by-swapping-elements/3219-make-lexicographically-smallest-array-by-swapping-elements.java
import java.util.*;

class Solution {
    public int[] lexicographicallySmallestArray(int[] nums, int limit) {
        int n = nums.length;
        List<int[]> help = new ArrayList<>();
        for (int i = 0; i < n; i++)
            help.add(new int[] {nums[i], i});
        
        help.sort((a, b) -> Integer.compare(a[0], b[0]));

        int[] res = new int[n];
        int prev = Integer.MIN_VALUE;
        List<Integer> pos = new ArrayList<>();
        
        for (int s = 0, e = 0; e < n;) {
            pos.add(help.get(e)[1]);
            prev = help.get(e)[0];
            e++;
            if (e == n || help.get(e)[0] - prev > limit) {
                Collections.sort(pos);
                for (int idx : pos) {
                    res[idx] = help.get(s)[0];
                    s++;
                }
                pos.clear();
            }
        }
        return res;
    }
}
